File: analysis/import/SCMLogParser/src/test/java/de/maibornwolff/codecharta/importer/scmlogparser/parser/LogLineCollectorTest.java
package de.maibornwolff.codecharta.importer.scmlogparser.parser;

import org.junit.Test;

import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collector;
import java.util.stream.Stream;

import static java.util.Collections.singletonList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class LogLineCollectorTest {

    @Test
    public void collectsBetweenSeparators() {
        Stream<String> logLines = Stream.of("##", "commit 1", "##", "commit 2");
        Collector<String, ?, Stream<List<String>>> collector = LogLineCollector.create(logLine -> logLine.equals("##"));
        Stream<List<String>> commits = logLines.collect(collector);
        assertThat(commits).containsExactly(singletonList("commit 1"), singletonList("commit 2"));
    }

    @Test
    public void raisesExceptionWhenNoCommitSeparatorPresent() {
        Stream<String> logLines = Stream.of("commit 1", "commit 2");
        Predicate<String> commitSeparatorTest = logLine -> logLine.equals("some commit separator");
        assertThatThrownBy(() -> logLines.collect(LogLineCollector.create(commitSeparatorTest))).isInstanceOf(IllegalArgumentException.class);
    }

    @Test
    public void skipsEmptyLines() {
        Stream<String> logLinesWithEmptyOneInTheMiddle = Stream.of("--", "commit data", "", "commit comment", "--");
        Collector<String, ?, Stream<List<String>>> collector = LogLineCollector.create(string -> string.equals("--"));
        Stream<List<String>> commits = logLinesWithEmptyOneInTheMiddle.collect(collector);
        assertThat(commits).hasSize(1);
    }

    @Test
    public void doesNotSupportParallelStreams() {
        Stream<String> logLines = Stream.of("--", "commit 1").parallel();
        Collector<String, ?, Stream<List<String>>> collector = LogLineCollector.create(logLine -> logLine.equals("--"));
        assertThatThrownBy(() -> logLines.collect(collector)).isInstanceOf(IllegalArgumentException.class);
    }
}

File: analysis/import/SCMLogParserV2/src/test/kotlin/de/maibornwolff/codecharta/importer/scmlogparserv2/parser/LogLineCollectorTest.kt
package de.maibornwolff.codecharta.importer.scmlogparserv2.parser

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.Test
import java.util.function.Predicate
import java.util.stream.Stream

class LogLineCollectorTest {

    @Test
    fun collectsBetweenSeparators() {
        val logLines = Stream.of("##", "commit 1", "##", "commit 2")
        val collector = LogLineCollector.create(Predicate { logLine -> logLine == "##" })
        val commits = logLines.collect(collector)
        assertThat(commits).containsExactly(listOf("commit 1"), listOf("commit 2"))
    }

    @Test
    fun raisesExceptionWhenNoCommitSeparatorPresent() {
        val logLines = Stream.of("commit 1", "commit 2")
        val commitSeparatorTest = Predicate<String> { logLine -> logLine == "some commit separator" }
        assertThatThrownBy { logLines.collect(LogLineCollector.create(commitSeparatorTest)) }.isInstanceOf(
            IllegalArgumentException::class.java
        )
    }

    @Test
    fun skipsEmptyLines() {
        val logLinesWithEmptyOneInTheMiddle = Stream.of("--", "commit data", "", "commit comment", "--")
        val collector = LogLineCollector.create(Predicate { string -> string == "--" })
        val commits = logLinesWithEmptyOneInTheMiddle.collect(collector)
        assertThat(commits).hasSize(1)
    }

    @Test
    fun doesNotSupportParallelStreams() {
        val logLines = Stream.of("--", "commit 1").parallel()
        val collector = LogLineCollector.create(Predicate { logLine -> logLine == "--" })
        assertThatThrownBy { logLines.collect(collector) }.isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun BOMIsRemoved() {
        val logLines = Stream.of("\uFEFF---", "whatever")
        val collector = LogLineCollector.create(Predicate { logLine -> logLine.startsWith("---") })
        val commits = logLines.collect(collector)
        assertThat(commits).hasSize(1)
    }
}
